proposals/004-routing-api.md

```diff
@@ -0,0 +1,248 @@
+
+```yaml
+routerDefinitions:
+  - name: my-router
+    type: MyRouter
+    config: ...
+    routes:
+      - name: foo
+        filters: 
+          - my-first-filter
+          - my-second-filter
+        cluster: my-backing-cluster
+      - name: bar
+        filters: 
+          - my-third-filter
+        router: my-other-router
+  - name: my-other-router
+    # ...
+```
+
+A route object has a `name`, an optional list of `filters` (being the names of the filters to be applied to requests/responses that traverse this route) and either a 
+`cluster` or a `router` property, which names the receiver which will handle requests after any filters have been applied. 
+Exactly one of `cluster` or `router` must be specified.
+
+Because routers can refer to other routers they form a graph. 
+
+```
+                           my-backing-cluster
+                          ^
+                         /
+                        / foo
+     requests          /
+    ---------> my-router
+                       \
+                        \ bar              
+                         \                ...
+                          v              /
+                           my-other-router --- ...
+                                         \
+                                          ...
+```
+
+All _possible_ routes through the graph can be determined statically from the proxy configuration, but the routing of any individual incoming request is determined at runtime
+can might involve multiple outgoing requests from any given router.
+Validation performed at proxy startup will reject cyclic graphs.
+This will prevent the possibility of a request getting stuck in a router loop. 
+
+In order for non-trivial router graphs to be useful, `Router` authors will need to follow the same principle of composition as `Filter` authors.
+That is, a `Router` implementation should only talk to its receivers, and not, for example, make direct connections of its own to a backing cluster.
+To do so would prevent use of that router implementation in a larger graph. 
+
+The `cluster` propety names a network-reachable backing cluster that speaks the Kafka procotol. It has the same schema as the `targetCluster` property of a virtual cluster.
+
+The existing virtual cluster schema will be modified to support top level `clusters` and to make use of `routers`.
+
+Specifically:
+
+* the existing `targetCluster` property will be made optional, and deprecated
+* a new `cluster` property will support referencing a target cluster by name (using a distict property name seems slightly nicer than overloading the allowed type of the existing `targetCluster` to support `string` or `object`).
+* support for new `router` property will be added. This is a reference to a router defined in `routerDefinitions`. 
+* exactly one of `router`, `cluster` or `targetCluster` will be required.
```

Comment:
Same comment as above, can there be a single property like `target` or `receiver` (living aside the aspect of deprecating the existing one), which can be of multiple types? Unless that's something discouraged in YAML, I'd prefer that, as it focuses on the role/purpose of the given entity in this context, rather than what it is by itself. Also, more potential kinds of targets can be added later on, without having an ever-growing list of mutually exclusive parameters.

Reply:
But ever-growing lists are the most exciting sort! :stuck_out_tongue_winking_eye: 

More seriously, let's keep discussion of this one in the previous thread.